### PR TITLE
Derive windows from SQANTI; update package structure

### PR DESCRIPTION
**Pull request by cmatKhan**
This is a relatively large change -- I am issuing the pull request b/c it is the easiest way to propose these changes. Note that there was no code changed in `scripts` only [src](https://github.com/cmatKhan/isocomp/tree/oop_ify/src/isocomp)

There are a few "package" items:

1. Main method has the parser/subparsers and entry points. This creates  cmd line interface like so:

```raw
(.venv) ➜  temp git:(oop_ify) isocomp --help
usage: isocomp [-h] [-v] {create_windows,find_unique_isoforms} ...

isocomp: 0.1.0

positional arguments:
  {create_windows,find_unique_isoforms}
                        Available Tools
    create_windows      from a list of SQANTI gtfs, create a bed6 formatted file -- 0 indexed, half open intervals -- where each entry in the bed file represents a discrete region of
                        coverage over the reference genome.
    find_unique_isoforms
                        Using the clustered gtf output of create_windows and a csv file with columns 'source' and 'fasta' which provide a map from the unique values in 'Source' in the
                        clustered gtf to the corresponding fasta file, find unique isoforms in the clusters described in the clustered_gtf.

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
```

the help for the two scripts are like so:

```raw
(.venv) ➜  temp git:(oop_ify) isocomp create_windows --help
usage: create_windows [-h] [-l {critical,error,warning,info,debug}] -i INPUT [INPUT ...] [-p OUTPUT_PREFIX] [--overwrite | --no-overwrite]

from a list of SQANTI gtfs, create a bed6 formatted file -- 0 indexed, half open intervals -- where each entry in the bed file represents a discrete region of coverage over the
reference genome.

optional arguments:
  -h, --help            show this help message and exit

general:
  -l {critical,error,warning,info,debug}, --log_level {critical,error,warning,info,debug}

input:
  -i INPUT [INPUT ...], --input INPUT [INPUT ...]
                        a space delimited list of paths to gtf files from a concatenated merged set of regions will be created

output:
  -p OUTPUT_PREFIX, --output_prefix OUTPUT_PREFIX
                        This should be either a file basename -- NO extension -- which will be written to the current working directory, or a valid path up to a filename, eg
                        /output/path/concat would result in the file output/path/concat.gtf being written
  --overwrite, --no-overwrite
                        By default, create_windows will overwrite a file with the same name in the output directory. Set --no-overwrite to avoid overwriting.
```

and

```raw
(.venv) ➜  temp git:(oop_ify) isocomp find_unique_isoforms --help
usage: find_unique_isoforms [-h] [-l {critical,error,warning,info,debug}] -a CLUSTERED_GTF -f FASTA_MAP [-p OUTPUT_PREFIX] [--overwrite | --no-overwrite]

Using the clustered gtf output of create_windows and a csv file with columns 'source' and 'fasta' which provide a map from the unique values in 'Source' in the clustered gtf to the
corresponding fasta file, find unique isoforms in the clusters described in the clustered_gtf.

optional arguments:
  -h, --help            show this help message and exit

general:
  -l {critical,error,warning,info,debug}, --log_level {critical,error,warning,info,debug}

input:
  -a CLUSTERED_GTF, --clustered_gtf CLUSTERED_GTF
                        path to the clustered gtf file, likely created by create_windows
  -f FASTA_MAP, --fasta_map FASTA_MAP
                        path to a csv file with columns 'source' and 'fasta' which provide a map between the unique factor levels of the Source column of the clustered gtf and a
                        corresponding fasta file which stores the transcript sequences

output:
  -p OUTPUT_PREFIX, --output_prefix OUTPUT_PREFIX
                        This should be either a file basename -- NO extension -- which will be written to the current working directory, or a valid path up to a filename, eg
                        /output/path/concat would result in the file output/path/concat.gtf being written
  --overwrite, --no-overwrite
                        By default, find_unique_isoforms will overwrite a file with the same name in the output directory. Set --no-overwrite to avoid overwriting.
```

All of the code in `src` is both documented/commented (this can always be improved, of course). The `mkdocs` documentation is also all set up for these objects with a little bit more added in -- if you `poetry install` to install the virtual environment, you can do `mkdocs serve` to serve the documentation and see what it is like.

2. Another feature is that the functions/classes in `src` have two methods of use: as cmd line scripts, and also interactively like an API. [The beginnings of a tutorial describing how to use these objects this way is here](https://github.com/cmatKhan/isocomp/blob/oop_ify/docs/api/api_tutorial.ipynb) (you can also see this in the docs if you `mkdocs serve`)  

3. The [test directory](https://github.com/cmatKhan/isocomp/tree/oop_ify/src/tests) is set up for all code in `src`. Files in the `test/data` directory are:

```raw
(.venv) ➜  temp git:(oop_ify) tree ../src/tests/data 
../src/tests/data
├── clustered_regions.gtf
├── fasta_map.csv
├── hg002_classification.txt
├── hg002_sqanti_fltr.fasta
├── hg002_sqanti_fltr.fasta.fai
├── hg002_sqanti_fltr.gtf
├── hg004_classification.txt
├── hg004_sqanti_fltr.fasta
├── hg004_sqanti_fltr.fasta.fai
├── hg004_sqanti_fltr.gtf
├── hg005_classification.txt
├── hg005_sqanti_fltr.fasta
├── hg005_sqanti_fltr.fasta.fai
├── hg005_sqanti_fltr.gtf
└── unique_isoforms.csv
```  
These are all small subsets, all of them created from the BCM SQANTI filtered files. the `classification` is from SQANTI and contains a good deal of additional annotation,etc that SQANTI thinks is interesting. `clustered_regions.gtf` is the output of `create_windows` and `unique_isoforms.csv` is the output of `find_unique_isoforms`. 
`fasta_map.csv` is an example fasta map, which is input to `find_unique_isoforms`.

4. As mentioned above, the mkdocs docs are more set up. Take a look with `mkdocs serve`.

5. All code in `src` has [tests](https://github.com/cmatKhan/isocomp/tree/oop_ify/src/tests) which at least act as easy entry points for further development.

# CAVEATS

1. The output of `find_unique_isoforms` is not intended for analysis, just for process. I didn't reference the output format, and in any case, I am doing this slightly differently by using the SQANTI data directly. Before I make this conform to the output standards, let's talk about if we like this method or not

2. There is a Dockerfile. I am at the in-laws on wireless, and it was taking a long time to build...so that is untested. if anyone has a better method of doing a Dockerfile, i'd be curious to see it.

### CODE REVIEW

**Comment by cmatKhan** [comment]
Sorry...issued the last one from my develop branch by mistake.

Tomorrow, headed out of Internet / cell range for a few days. Back on the third unless the chupa cabra gets us.

**Comment by cmatKhan** [comment]
create_windows wasn't working due to a dependency error, but that is now resolved, thanks to NASA:

https://github.com/nasa/puma/issues/9

File: scripts/find_unique_isoform.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import edlib
 import numpy
 import logging
+import logging
```
Comment (ame9yu) [inline scripts/find_unique_isoform.py:11]: There are a lot of repeated lines. This might be due to an automatic merging. Could you resolve this?

Comment (cmatKhan) [reply scripts/find_unique_isoform.py:11]: hmmmm...yes, probably true.

I didn't intentionally change anything in `scripts`, but did do some merging with `main` and whatnot -- i must have messed something up there.

If this pull is going to be accepted, then anything in `scripts` upstream of finding unique isoforms is obsolete. For downstream functions, I'll make sure that I have an up-to-date and correct version, and then integrate it into `src` and the cmd line interface before the pull is accepted/merged.

**Comment by cmatKhan** [comment]
Per a question from yutong on slack, here is a suggestion of how to look at this code:

The big change is how the "bins" are formed -- I am using the SQANTI GTF files.  

After that, the TX comparisons aren't intended to be different than the current implementation (though, as you said, written in a more maintenance friendly way). That said, I didn't put any effort into conforming to the 'output' format -- don't want to do that unless there is consensus on this method.For code review purposes, I'd suggest this:

1.  If you want to just set some breakpoints and see what the code does, pull my [oop_ify](https://github.com/cmatKhan/isocomp/tree/oop_ify) branch. poetry install the environment and then run the tests. You can set breakpoints in the src directory code files if you want to stop and take a look around.
1. If you just want to read, for the create_windows, [read this](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Coordinates/create_comparison_windows.py). The docstring should be clear. It might be helpful to read about what pyranges [concat](https://biocore-ntnu.github.io/pyranges/concatenating-pyranges.html) and [cluster](https://biocore-ntnu.github.io/pyranges/methods-for-manipulating-single-pyranges.html) do. Example of the concatted, clustered gtf is [here](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/tests/data/clustered_regions.gtf).
1. the 'main' method for comparison within bins is [here](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Compare/find_unique_isoforms.py). In the loop, `for cluster in il.cluster_list`, it just loops over the clusters created by pyranges cluster (see above). `il` is a [IsoformLibrary](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Compare/IsoformLibrary.py) object (I added two classes, this one and [Window](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Coordinates/Window.py)). Each object or function have their own file, so you can follow the logic in find_unique_isoforms by looking at the functions/objects which it calls, most of which are in the [Compare](https://github.com/cmatKhan/isocomp/tree/oop_ify/src/isocomp/Compare) module.

If you are quite familiar with the argparse module, you could just start on the isocomp [__main__](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/__main__.py) and follow through the code from there.

**Comment by ame9yu** [comment]
Could you explain why this change is made and how the change affects the overall efficiency and the performance of the current software? 

> Per a question from yutong on slack, here is a suggestion of how to look at this code:
> 
> The big change is how the "bins" are formed -- I am using the SQANTI GTF files.
> 
> After that, the TX comparisons aren't intended to be different than the current implementation (though, as you said, written in a more maintenance friendly way). That said, I didn't put any effort into conforming to the 'output' format -- don't want to do that unless there is consensus on this method.For code review purposes, I'd suggest this:
> 
> 1. If you want to just set some breakpoints and see what the code does, pull my [oop_ify](https://github.com/cmatKhan/isocomp/tree/oop_ify) branch. poetry install the environment and then run the tests. You can set breakpoints in the src directory code files if you want to stop and take a look around.
> 2. If you just want to read, for the create_windows, [read this](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Coordinates/create_comparison_windows.py). The docstring should be clear. It might be helpful to read about what pyranges [concat](https://biocore-ntnu.github.io/pyranges/concatenating-pyranges.html) and [cluster](https://biocore-ntnu.github.io/pyranges/methods-for-manipulating-single-pyranges.html) do. Example of the concatted, clustered gtf is [here](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/tests/data/clustered_regions.gtf).
> 3. the 'main' method for comparison within bins is [here](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Compare/find_unique_isoforms.py). In the loop, `for cluster in il.cluster_list`, it just loops over the clusters created by pyranges cluster (see above). `il` is a [IsoformLibrary](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Compare/IsoformLibrary.py) object (I added two classes, this one and [Window](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/Coordinates/Window.py)). Each object or function have their own file, so you can follow the logic in find_unique_isoforms by looking at the functions/objects which it calls, most of which are in the [Compare](https://github.com/cmatKhan/isocomp/tree/oop_ify/src/isocomp/Compare) module.
> 
> If you are quite familiar with the argparse module, you could just start on the isocomp [**main**](https://github.com/cmatKhan/isocomp/blob/oop_ify/src/isocomp/__main__.py) and follow through the code from there.

**Comment by cmatKhan** [comment]
We are taking assembled TX as input to isocomp. SQANTI, at least, and likely other TX assemblers, use the original alignments _and_ the reference genome to aid in TX assembly and error correction. The output of SQANTI, and I am guessing other TX assemblers, is then a fasta file of TX sequences _and_ a GTF file. 

The GTF file fully describes the coordinates of the assembled TX on the reference genome which was used to assemble the TX.

What I am proposing is to eliminate the step which aligns the TX back to the genome (again), and to use the assembler's GTF file to create the comparison windows ([see number 2 above](https://github.com/collaborativebioinformatics/isocomp/pull/31#issuecomment-1378168477)). By concatting any number of SQANTI GTFs, then merging (I'm using the pyranges cluster function, which is a variant of merge) regions, this divides the genome into bins which fully capture the transcripts in the fasta. **This simplifies the code in isocomp, and avoids some questions about the alignment method which I think are thorny**. 

In comparing the alignment vs GTF method, there are two cases:

1. The aligner and GTF agree, in which case the additional time to align is unnecessary, and the additional code in isocomp is just a maintenance burden
2. The aligner and GTF disagree, and then we have to ask the question, "are we going to re-assemble the transcript or throw it out?" Re-assembly is out of the scope of isocomp. Throwing it out is a possibility.

Additionally, SQANTI already classifies the assembled isoforms according to how they compare to the reference genome which is used to assemble/error corect. The SQANTI paper fig 1 describes this:

https://pubmed.ncbi.nlm.nih.gov/29440222/#&gid=article-figures&pid=figure-1-uid-0

In the code I am proposing, I am making it simple to join the Isocomp output back to the SQANTI additional annotations (no renaming -- I can explain this in a meeting later if we are going to go ahead with this).

One benefit is that this simplifies the "fusion" gene question -- in the filtered HG002, for instance, SQANTI identifies 161 fusion genes. From what Evan has been saying, there are likely few false negatives and most of these 161 are false positives. Isocomp would provide additional tools to evaluate and filter these fusion calls.

Other items in the code of this proposal are more aesthetic and/or package related. 

1. The functions are broken out into their own files, so it is easier to find and focus on just the "alignment" function, for instance. This improves readability (though arguably it is harder to understand how each function is intended to fit together -- that is a documentation issue, though), and greatly improves test-ability -- it is a lot easier to test short, limited purpose functions. To this end, I made an effort to keep each function short and specific to its task, and the docstrings, comments and variables are intended to be easily readable.
2. The separation of the code into discrete functions/classes makes it easier to re-use code, and also to use each one of these functions/classes in a notebook/interactive session
3. Related to 2, parallelizing the in-window sequence comparisons across windows on a HPC would be trivial in this code. Further parallelizing the sequence comparisons themselves would probably also be worthwhile -- also an easy improvement to this code.
4. I do my best to follow the [python pep8 style guide](https://peps.python.org/pep-0008/) . The first paragraph in the PEP8 does note that blind adherence to a coding standard is small minded. That said, it really helps in group projects and for long term maintenance to adhere to a set of standards, and these are commonly used.
5. For all code in `src`, there exists a test that at least allows easy access to the code. Test data is also provided.
6. the mkdocs documentation is set up to read the documentation from the docstrings automatically, though admittedly the formatting leaves something to be desired. This is definitely an area that can be iteratively improved. Other parts of the documentation have been built out, too, at least in a skeleton form.
7. Entry points (cmd line scripts) are configured in `__main__`, so when a user installs isocomp, they have access to the cmd `isocomp` on the cmd line. They could also run it with `python -m isocomp`, which is sort of nice if you want to be very sure about what python you are using.

**Comment by evanbiederstedt** [comment]
(Just a note of apologies that I haven't looked at this properly yet; I'm still drowning a bit; please yell at me on slack)

**Comment by evanbiederstedt** [comment]
I should be able to return to this shortly. Apologies again to @cmatKhan that I've ignored this for so long. 

If you have time, please review and ask question @ame9yu @BidaGu 

Perhaps the next step is for Chase to record a short video (similar to Zoom presentations whereby the speaker shares screen) and walks through the code for us. A recorded walkthrough will be better than a meeting I think (given everyone's schedules and timezone differences). 

What do you think @cmatKhan ?

**Comment by cmatKhan** [comment]
videos posted to slack

**Comment by cmatKhan** [comment]
i just made some adjustments based on the last meeting, and a discussion on slack -- see the "upcoming notes" as of 20230417 in the isocomp shared google docs -- in this commit: [3708f43](https://github.com/collaborativebioinformatics/isocomp/pull/31/commits/3708f43accf1dc2b925ee7479a32a433d04dbfd4)

It is worth looking at this:

[Beginnings of the notebook documenting the API](https://github.com/cmatKhan/isocomp/blob/oop_ify/docs/api/api_tutorial.ipynb)  -- __scroll down to the bottom__

where i tried to think of the different ways that a set of isoforms might form a cluster (or window -- our comparison bins).

The blue lines are supposed to be isoform features, and the red dotted line(s) are the clusters they form. Note that the blue lines could come from 3 different subjects, or entirely the same subject, or any combination.

**Review by evanbiederstedt — APPROVED** [review]
LGTM

Thanks @cmatKhan !!

**Pull request merged**